core/io/image.cpp

```diff
@@ -3929,22 +3929,78 @@ void Image::renormalize_rgbe9995(uint32_t *p_rgb) {
+	}
+
+	static const size_t riff_four_cc_size = 4;
+	static const uint8_t riff_file_header[] = { 0x52, 0x49, 0x46, 0x46 }; // RIFF
+	static const uint8_t webp_content_header[] = { 0x57, 0x45, 0x42, 0x50 }; // WEBP
+	if (len > riff_four_cc_size * 3 && // RIFF SIZE WEBP
+			memcmp(p_data, riff_file_header, riff_four_cc_size) == 0 && // RIFF header.
+			memcmp(p_data + riff_four_cc_size * 2, webp_content_header, riff_four_cc_size) == 0) { // WEBP chunk header.
+		return FILE_FORMAT_RIFF_WEBP;
+	}
+
+	static const uint8_t jpeg_soi_marker[] = { 0xFF, 0xD8 };
+	static const uint8_t jpeg_application0_marker[] = { 0xFF, 0xE0 };
+	static const uint8_t jpeg_jfif_data_identifier[] = { 0x4A, 0x46, 0x49, 0x46, 0x00 }; // JFIF\0
+	static const uint8_t jpeg_application1_marker[] = { 0xFF, 0xE1 };
+	static const uint8_t jpeg_exif_data_identifier[] = { 0x45, 0x78, 0x69, 0x66, 0x00, 0x00 }; // Exif\0\0
+
+	static const size_t jpeg_marker_size = 2;
+	static const size_t jpeg_data_length_size = sizeof(uint16_t);
+	static const size_t jpeg_data_identifier_offset = jpeg_marker_size * 2 + jpeg_data_length_size;
+	static const size_t jpeg_jfif_data_identifier_size = sizeof(jpeg_jfif_data_identifier);
+	static const size_t jpeg_exif_data_identifier_size = sizeof(jpeg_exif_data_identifier);
+
+	if (len > 12 && memcmp(p_data, jpeg_soi_marker, jpeg_marker_size) == 0) {
+		if (memcmp(p_data + jpeg_marker_size, jpeg_application0_marker, jpeg_marker_size) == 0 &&
+				memcmp(p_data + jpeg_data_identifier_offset,
+						jpeg_jfif_data_identifier, jpeg_jfif_data_identifier_size) == 0) {
+			return FILE_FORMAT_JPEG_JFIF;
+		} else if (memcmp(p_data + jpeg_marker_size, jpeg_application1_marker, jpeg_marker_size) == 0 &&
+				memcmp(p_data + jpeg_data_identifier_offset,
+						jpeg_exif_data_identifier, jpeg_exif_data_identifier_size) == 0) {
+			return FILE_FORMAT_JPEG_EXIF;
+		}
+	}
+
+	return FILE_FORMAT_UNKNOWN;
+}
+
 Image::Image(const uint8_t *p_mem_png_jpg, int p_len) {
 	width = 0;
 	height = 0;
 	mipmaps = false;
 	format = FORMAT_L8;
 
-	if (_png_mem_loader_func) {
-		copy_internals_from(_png_mem_loader_func(p_mem_png_jpg, p_len));
-	}
-
-	if (is_empty() && _jpg_mem_loader_func) {
-		copy_internals_from(_jpg_mem_loader_func(p_mem_png_jpg, p_len));
-	}
-
-	if (is_empty() && _webp_mem_loader_func) {
-		copy_internals_from(_webp_mem_loader_func(p_mem_png_jpg, p_len));
+	switch (detect_file_format(p_mem_png_jpg, p_len)) {
+		case FILE_FORMAT_PNG:
+			if (_png_mem_loader_func) {
+				copy_internals_from(_png_mem_loader_func(p_mem_png_jpg, p_len));
+			}
+			break;
+		case FILE_FORMAT_JPEG_JFIF:
+		case FILE_FORMAT_JPEG_EXIF:
+			if (_jpg_mem_loader_func) {
+				copy_internals_from(_jpg_mem_loader_func(p_mem_png_jpg, p_len));
+			}
+			break;
+		case FILE_FORMAT_RIFF_WEBP:
+			if (_webp_mem_loader_func) {
+				copy_internals_from(_webp_mem_loader_func(p_mem_png_jpg, p_len));
+			}
+			break;
+		default: // FILE_FORMAT_UNKNOWN
+			initialize_data(1, 1, false, FORMAT_RGBA8);
```

Comment:
During the checks, an interesting phenomenon was discovered: all the checks pass except for one.

It seems that in the assembly for Linux, an icon in an invalid format is loaded into the DisplayServerX11. Regardless, the image is created. I can't find out what the file is and I can't check what image was created from it. But it certainly turns out to be non-empty.

Adding my format checks caused this image to stop being generated (had a width and a height, but didn't have a single byte in the data block). This caused the application to crash in this test.

That's why I put a 1 pixel transparent image initialization here to avoid such errors.

Here is a listing of the log so that you can find the damaged image:

### 🐧 Linux / Editor with doubles and GCC sanitizers (target=editor, tests=yes, dev_build=yes, precision=double, use_asan=yes, use_ubsan=yes, linker=gold)

```
ERROR: FATAL: Index p_index = 2 is out of bounds (((Vector<T> *)(this))->_cowdata.size() = 2).
   at: operator[] (./core/templates/vector.h:52)

================================================================
handle_crash: Program crashed with signal 4
Engine version: Godot Engine v4.1.dev.custom_build (db8d3213c200ac922dd7ad0199127b13d674d006)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] ./bin/godot.linuxbsd.editor.dev.double.x86_64.san(+0x3cb0bac4) [0x55c042a3eac4] (??:?)
[2] /lib/x86_64-linux-gnu/libc.so.6(+0x43090) [0x7efed6d93090] (??:0)
[3] VectorWriteProxy<long>::operator[](int) (??:?)
[4] DisplayServerX11::set_icon(Ref<Image> const&) (??:?)
[5] Main::start() (??:?)
[6] ./bin/godot.linuxbsd.editor.dev.double.x86_64.san(main+0x3da) [0x55c042a3e1c3] (??:?)
[7] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7efed6d74083] (??:0)
[8] ./bin/godot.linuxbsd.editor.dev.double.x86_64.san(_start+0x2e) [0x55c042a3dd2e] (??:?)
-- END OF BACKTRACE --
================================================================
Aborted (core dumped)
FATAL ERROR: Godot has been crashed.
Error: Process completed with exit code 52.
```